p2p/protocol.go

```diff
@@ -14,39 +15,72 @@ import (
 
 type MessageType uint32
 
+type Item struct {
+	id        uint64
+	timestamp time.Time
+}
+
 type Protocol struct {
-	count              uint32
+	id                 uint64
 	name               string
 	network            Service
 	pendMutex          sync.RWMutex
-	pending            map[uint32]chan interface{}
-	resHandlers        map[uint32]func(msg []byte)
+	pendingMap         map[uint64]chan interface{}
+	pendingQueue       *list.List
+	resHandlers        map[uint64]func(msg []byte)
 	msgRequestHandlers map[MessageType]func(msg []byte) []byte
 	ingressChannel     chan service.Message
+	requestTimeout     time.Duration
+	Exit               chan struct{}
 }
 
-func NewProtocol(network Service, name string) *Protocol {
+func NewProtocol(network Service, name string, requestTimeout time.Duration) *Protocol {
 	p := &Protocol{
 		name:               name,
-		pending:            make(map[uint32]chan interface{}),
-		resHandlers:        make(map[uint32]func(msg []byte)),
 		network:            network,
+		requestTimeout:     requestTimeout,
+		pendingQueue:       list.New(),
 		ingressChannel:     network.RegisterProtocol(name),
+		pendingMap:         make(map[uint64]chan interface{}),
+		resHandlers:        make(map[uint64]func(msg []byte)),
 		msgRequestHandlers: make(map[MessageType]func(msg []byte) []byte),
+		Exit:               make(chan struct{}),
```

Comment:
never used, stop readloop on this. if this is exported then it preferred to be read/write only. the convention would be to close it in a `Close` method.